src/bitmap/store/bitmap_store.rs

```diff
@@ -309,10 +317,14 @@ impl<B: Borrow<[u64; BITMAP_LENGTH]>> Iterator for BitmapIter<B> {
         loop {
             if self.value == 0 {
                 self.key += 1;
-                if self.key >= BITMAP_LENGTH {
+                if self.key > self.key_back {
                     return None;
                 }
-                self.value = unsafe { *self.bits.borrow().get_unchecked(self.key) };
+                self.value = if self.key == self.key_back {
+                    self.value_back
+                } else {
+                    unsafe { *self.bits.borrow().get_unchecked(self.key) }
+                };
```

Comment:
Surprisingly: This was _faster_ to iterate for `into_iter` but slower for `iter`

Data:
```
group                                         double-ended-iter-stable                iter-head-stable
-----                                         ------------------------                ----------------
iteration/into_iter/census-income             1.00     40.9±0.02ms 161.5 MElem/sec    1.12     45.8±0.01ms 144.3 MElem/sec
iteration/into_iter/census-income_srt         1.00     35.6±0.00ms 163.4 MElem/sec    1.13     40.0±0.01ms 145.1 MElem/sec
iteration/into_iter/census1881                1.00      5.5±0.00ms 173.0 MElem/sec    1.00      5.6±0.00ms 172.3 MElem/sec
iteration/into_iter/census1881_srt            1.00      3.6±0.00ms 180.0 MElem/sec    1.18      4.2±0.00ms 153.0 MElem/sec
iteration/into_iter/weather_sept_85           1.00     77.6±0.04ms 158.2 MElem/sec    1.10     85.4±0.07ms 143.7 MElem/sec
iteration/into_iter/weather_sept_85_srt       1.00     94.2±0.01ms 163.1 MElem/sec    1.13    106.2±0.05ms 144.6 MElem/sec
iteration/into_iter/wikileaks-noquotes        1.00  1447.7±44.04µs 181.4 MElem/sec    1.10  1588.4±93.39µs 165.3 MElem/sec
iteration/into_iter/wikileaks-noquotes_srt    1.00  1567.8±48.13µs 175.2 MElem/sec    1.15  1808.2±16.45µs 151.9 MElem/sec
iteration/iter/census-income                  1.05     44.9±0.06ms 146.9 MElem/sec    1.00     42.6±0.23ms 154.8 MElem/sec
iteration/iter/census-income_srt              1.05     39.2±0.04ms 148.3 MElem/sec    1.00     37.3±0.27ms 155.9 MElem/sec
iteration/iter/census1881                     1.14      6.0±0.00ms 158.9 MElem/sec    1.00      5.3±0.03ms 180.6 MElem/sec
iteration/iter/census1881_srt                 1.06      4.3±0.00ms 150.9 MElem/sec    1.00      4.1±0.02ms 160.1 MElem/sec
iteration/iter/weather_sept_85                1.06     84.2±0.11ms 145.7 MElem/sec    1.00     79.6±0.38ms 154.3 MElem/sec
iteration/iter/weather_sept_85_srt            1.04    103.2±0.12ms 148.9 MElem/sec    1.00     99.0±0.61ms 155.2 MElem/sec
iteration/iter/wikileaks-noquotes             1.14   1680.4±0.41µs 156.3 MElem/sec    1.00  1472.1±12.32µs 178.4 MElem/sec
iteration/iter/wikileaks-noquotes_srt         1.07   1830.2±0.53µs 150.1 MElem/sec    1.00  1705.3±12.77µs 161.1 MElem/sec
```

We can do better still! Github wont let me apply a suggestion here because this section contains deleted lines, but here it is

```rust
impl<B: Borrow<[u64; BITMAP_LENGTH]>> Iterator for BitmapIter<B> {
    type Item = u16;

    fn next(&mut self) -> Option<u16> {
        loop {
            if self.value == 0 {
                self.key += 1;
                // Match arms can be reordered, this ordering is perf sensitive
                let cmp = self.key.cmp(&self.key_back);
                self.value = if cmp == Ordering::Less {
                    unsafe { *self.bits.borrow().get_unchecked(self.key) }
                } else if cmp == Ordering::Equal {
                    self.value_back
                } else {
                    return None;
                };
                continue;
            }
            let index = self.value.trailing_zeros() as usize;
            self.value &= self.value - 1;
            return Some((64 * self.key + index) as u16);
        }
    }
}
```


Data:
group                                         double-ended-iter-stable                double-ended-iter-stable-saik0
-----                                         ------------------------                ------------------------------
iteration/into_iter/census-income             1.04     40.9±0.02ms 161.5 MElem/sec    1.00     39.4±0.01ms 167.6 MElem/sec
iteration/into_iter/census-income_srt         1.04     35.6±0.00ms 163.4 MElem/sec    1.00     34.1±0.01ms 170.3 MElem/sec
iteration/into_iter/census1881                1.00      5.5±0.00ms 173.0 MElem/sec    1.00      5.5±0.00ms 173.5 MElem/sec
iteration/into_iter/census1881_srt            1.03      3.6±0.00ms 180.0 MElem/sec    1.00      3.5±0.00ms 186.1 MElem/sec
iteration/into_iter/weather_sept_85           1.03     77.6±0.04ms 158.2 MElem/sec    1.00     75.5±0.05ms 162.6 MElem/sec
iteration/into_iter/weather_sept_85_srt       1.04     94.2±0.01ms 163.1 MElem/sec    1.00     90.4±0.02ms 170.0 MElem/sec
iteration/into_iter/wikileaks-noquotes        1.00  1447.7±44.04µs 181.4 MElem/sec    1.00  1452.5±42.53µs 180.8 MElem/sec
iteration/into_iter/wikileaks-noquotes_srt    1.03  1567.8±48.13µs 175.2 MElem/sec    1.00  1527.1±46.15µs 179.9 MElem/sec
iteration/iter/census-income                  1.08     44.9±0.06ms 146.9 MElem/sec    1.00     41.6±0.04ms 158.8 MElem/sec
iteration/iter/census-income_srt              1.09     39.2±0.04ms 148.3 MElem/sec    1.00     36.1±0.04ms 161.2 MElem/sec
iteration/iter/census1881                     1.00      6.0±0.00ms 158.9 MElem/sec    1.00      6.0±0.00ms 159.4 MElem/sec
iteration/iter/census1881_srt                 1.06      4.3±0.00ms 150.9 MElem/sec    1.00      4.1±0.00ms 160.2 MElem/sec
iteration/iter/weather_sept_85                1.06     84.2±0.11ms 145.7 MElem/sec    1.00     79.2±0.05ms 154.9 MElem/sec
iteration/iter/weather_sept_85_srt            1.08    103.2±0.12ms 148.9 MElem/sec    1.00     95.3±0.07ms 161.2 MElem/sec
iteration/iter/wikileaks-noquotes             1.00   1680.4±0.41µs 156.3 MElem/sec    1.00   1679.7±0.67µs 156.3 MElem/sec
iteration/iter/wikileaks-noquotes_srt         1.05   1830.2±0.53µs 150.1 MElem/sec    1.00   1740.7±1.09µs 157.8 MElem/sec

I'm happy with the perf characteristics of `DoubleEndedIter` compared to a specialized `iter_rev`. Lets go with this route.
```

Reply:
Nice! Made this change